File: algorithmic-thinking/search-algorithms/linear-search/linear_search.py
# 线性搜索 - Python 实现

"""
线性搜索算法
时间复杂度: O(n)
空间复杂度: O(1)
"""

def linear_search(arr, target):
    """
    线性搜索：逐个遍历数组查找目标值
    返回目标值的索引，未找到返回 -1
    """
    for i in range(len(arr)):
        if arr[i] == target:
            return i
    return -1

def linear_search_all(arr, target):
    """
    查找数组中所有等于目标值的索引
    返回索引列表
    """
    indices = []
    for i in range(len(arr)):
        if arr[i] == target:
            indices.append(i)
    return indices

def linear_search_with_condition(arr, condition):
    """
    根据条件函数查找第一个满足条件的元素
    condition: 接收元素，返回 True/False 的函数
    """
    for i in range(len(arr)):
        if condition(arr[i]):
            return i
    return -1

if __name__ == "__main__":
    print("=" * 50)
    print("线性搜索 - Python 实现")
    print("=" * 50)
    
    # 测试用例 1
    print("\n测试 1: 查找单个元素")
    arr = [5, 2, 8, 1, 9, 3, 7]
    target = 8
    result = linear_search(arr, target)
    print(f"数组: {arr}")
    print(f"查找: {target}")
    print(f"结果: {result}")
    
    # 测试用例 2
    print("\n测试 2: 查找不存在的元素")
    target = 10
    result = linear_search(arr, target)
    print(f"查找: {target}")
    print(f"结果: {result}")
    
    # 测试用例 3
    print("\n测试 3: 查找所有匹配元素")
    arr = [1, 2, 3, 2, 4, 2, 5]
    target = 2
    results = linear_search_all(arr, target)
    print(f"数组: {arr}")
    print(f"查找所有 {target}: {results}")
    
    # 测试用例 4
    print("\n测试 4: 条件查找")
    arr = [1, 2, 3, 4, 5, 6, 7, 8]
    result = linear_search_with_condition(arr, lambda x: x > 5)
    print(f"数组: {arr}")
    print(f"查找第一个 > 5 的元素")
    print(f"结果: {result} (值: {arr[result] if result != -1 else '未找到'})")
    
    print("\n" + "=" * 50)

File: algorithmic-thinking/search-algorithms/linear-search/linear_search.rs
/// 线性搜索（Linear Search）- Rust 版本
///
/// 依次扫描切片中的元素，找到目标值所在的下标。
/// 时间复杂度：O(n)
/// 空间复杂度：O(1)

/// 返回第一个等于 target 的下标，未找到返回 None
pub fn linear_search<T: PartialEq>(arr: &[T], target: &T) -> Option<usize> {
    for (i, v) in arr.iter().enumerate() {
        if v == target {
            return Some(i);
        }
    }
    None
}

/// 返回所有等于 target 的下标
pub fn linear_search_all<T: PartialEq>(arr: &[T], target: &T) -> Vec<usize> {
    let mut indices = Vec::new();
    for (i, v) in arr.iter().enumerate() {
        if v == target {
            indices.push(i);
        }
    }
    indices
}

fn main() {
    println!("========== Linear Search (Rust) ==========");
    let arr = vec![5, 2, 8, 1, 9, 3, 7];
    println!("数组: {:?}", arr);
    println!("查找 8 -> {:?}", linear_search(&arr, &8));
    println!("查找 10 -> {:?}", linear_search(&arr, &10));

    let arr2 = vec![1, 2, 3, 2, 4, 2];
    println!("数组2: {:?}", arr2);
    let indices = linear_search_all(&arr2, &2);
    println!("查找所有 2 -> {:?}", indices);
}

